src/main/java/org/opensearch/knn/index/quantizationService/KNNVectorQuantizationTrainingRequest.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.quantizationService;
+
+import org.opensearch.knn.index.vectorvalues.KNNVectorValues;
+import org.opensearch.knn.quantization.models.requests.TrainingRequest;
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+
+/**
+ * KNNVectorQuantizationTrainingRequest is a concrete implementation of the abstract TrainingRequest class.
+ * It provides a mechanism to retrieve float vectors from the KNNVectorValues by document ID.
+ */
+final class KNNVectorQuantizationTrainingRequest<T> extends TrainingRequest<T> {
+
+    private final KNNVectorValues<T> knnVectorValues;
+    private int lastIndex;
+
+    /**
+     * Constructs a new QuantizationFloatVectorTrainingRequest.
+     *
+     * @param knnVectorValues the KNNVectorValues instance containing the vectors.
+     */
+    KNNVectorQuantizationTrainingRequest(KNNVectorValues<T> knnVectorValues) {
+        super((int) knnVectorValues.totalLiveDocs());
+        this.knnVectorValues = knnVectorValues;
+        this.lastIndex = 0;
+    }
+
+    /**
+     * Retrieves the float vector associated with the specified document ID.
+     *
+     * @param docId the document ID.
+     * @return the float vector corresponding to the specified document ID, or null if the docId is invalid.
+     */
+    @Override
+    public T getVectorByDocId(int docId) {
+        try {
+            while (lastIndex <= docId) {
+                knnVectorValues.nextDoc();
+                lastIndex++;
+            }
+            if (knnVectorValues.docId() == NO_MORE_DOCS) {
+                return null;
+            }
```

Comment:
isn't this an exception scenario, if current docId is NO_MORE_DOCS it means the number of vectors provided to sampler was wrong and the whole quantization should fail rather than returning null.

Reply:
This is getting handled while Calculating thresholds, if vector is null , it will throw exception.

Reply:
Here is the thing:
You function has 3 behaviors:
1. Return a valid vector.
2. Return a null vector
3. Throw exception.

You can make it just 2 and free the caller from knowing that vector could be null. This will reduce the branches on caller and on this function and make the code simpler. 

Also, this logic then will not need java docs on why we are adding it and any other conditions
```
if (knnVectorValues.docId() == NO_MORE_DOCS) {
                return null;
            }
```

Reply:
first and two is same ..which is return vector ..it could be null or anything else...

Reply:
this is what any function does ...throwing exception from function is pretty normal thing.

Reply:
> first and two is same ..which is return vector ..it could be null or anything else...

This is the reason I said returns a valid Vector. Null is not a valid vector. Returning null when we already know that it is not the right behavior for caller why don't just throw the exception. I am not able to understand the downside here.

Reply:
I have removed any catching the exception and throwing and relying on upstream, right now this function simple enough to return vector or null , any how when I throw any expception from here ..Client has to know about that ..I will let Clinet handle Null and take call appropriately . I am closing this , If you have any further improvement I can take in  follow up PR.